control/controller.go

```diff
@@ -84,228 +69,202 @@ func New(c Config) *Controller {
+	}
 
-		// Peek at head of priority queue
-		q := pq.Peek()
-		if q != nil {
-			pop, err := c.processQuery(q)
-			if pop {
-				pq.Pop()
-			}
-			if err != nil {
-				q.setErr(err)
-			}
-		}
+	// Lock the queries mutex for the rest of this method.
+	c.queriesMu.Lock()
+	defer c.queriesMu.Unlock()
+
+	if c.shutdown {
+		// Query controller was shutdown between when we started
+		// creating the query and ending it.
+		err := errors.New("query controller shutdown")
+		q.setErr(err)
+		return nil, err
 	}
+	c.queries[id] = q
+	return q, nil
 }
 
-// drain will continue processing queries from the priority queue and
-// processing done queries.
-func (c *Controller) drain(pq *PriorityQueue) {
-	for {
-		c.queriesMu.RLock()
-		if len(c.queries) == 0 {
-			c.queriesMu.RUnlock()
-			return
-		}
-		c.queriesMu.RUnlock()
+func (c *Controller) nextID() QueryID {
+	nextID := atomic.AddUint64(&c.lastID, 1)
+	return QueryID(nextID)
+}
 
-		// Wait for resources to free
-		q := <-c.queryDone
-		c.free(q)
-		c.queriesMu.Lock()
-		delete(c.queries, q.id)
-		c.queriesMu.Unlock()
+func (c *Controller) countQueryRequest(q *Query, result requestsLabel) {
+	l := len(q.labelValues)
+	lvs := make([]string, l+1)
+	copy(lvs, q.labelValues)
+	lvs[l] = string(result)
+	c.metrics.requests.WithLabelValues(lvs...).Inc()
+}
 
-		// Peek at head of priority queue
-		q = pq.Peek()
-		if q != nil {
-			pop, err := c.processQuery(q)
-			if pop {
-				pq.Pop()
-			}
-			if err != nil {
-				go q.setErr(err)
-			}
-		}
+func (c *Controller) compileQuery(q *Query, compiler flux.Compiler) error {
+	if !q.tryCompile() {
+		return errors.New("failed to transition query to compiling state")
 	}
-}
 
-// processQuery move the query through the state machine and returns and errors and if the query should be popped.
-func (c *Controller) processQuery(q *Query) (pop bool, err error) {
-	defer func() {
-		if e := recover(); e != nil {
-			// If a query panicked, always pop it from the queue so we don't
-			// try to reprocess it.
-			pop = true
-
-			// Update the error with information about the query if this is an
-			// error type and create an error if it isn't.
-			switch e := e.(type) {
-			case error:
-				err = errors.Wrap(e, "panic")
-			default:
-				err = fmt.Errorf("panic: %s", e)
-			}
-			if entry := c.logger.Check(zapcore.InfoLevel, "Controller panic"); entry != nil {
-				entry.Stack = string(debug.Stack())
-				entry.Write(zap.Error(err))
-			}
-		}
-	}()
+	prog, err := compiler.Compile(q.currentCtx)
+	if err != nil {
+		return errors.Wrap(err, "compilation failed")
+	}
+	q.program = prog
+	return nil
+}
 
-	// Check if we have enough resources
-	if c.check(q) {
-		// Update resource gauges
-		c.consume(q)
+func (c *Controller) enqueueQuery(q *Query) error {
+	if !q.tryQueue() {
+		return errors.New("failed to transition query to queueing state")
+	}
 
-		// Remove the query from the queue
-		pop = true
+	// TODO(jsternberg): We should have a real queue! It should only
```

Comment:
Why don't we keep the code for the real queue and we use it when switching to execute?
This would require minimum execution overhead, but would prevent us to reimplement the queue logic in a second time.

Reply:
The code for the real queue is something we probably want to redesign so I thought it would be too complicated to keep it. Most of this logical is about resource checking which we aren't keeping.

The new queue is probably going to be "check if we have the minimum resources for a task, pop off the first element in the queue, then start it. block if there aren't enough resources." We might decide to do more, but that's even further in the future as there's no designed way for us to request resource allocations or limits yet.